File: cmd/control-api/svcjournal.go
package main

// HEALTH-005 — the service journal, written from ONE place.
//
// Every API call passes through `guard` (access.go), which already knows the route's declaration, the
// caller, and whether the row being touched is theirs. So the journal hangs there rather than in
// forty-two handlers: a route added tomorrow is recorded because it is in the routes table, not
// because somebody remembered. That is the same reasoning ADR-109 used for the credential checks, and
// for the same reason — thirteen hand-placed copies are how the defect it fixed was produced.
//
// LEVELS DO THE WORK THAT SELECTION WOULD OTHERWISE DO (Alex's directive: record everything, and have
// levels). Reads are `debug`; the hub polls `/v1/runs/{id}` every 2s, so a ten-minute run is ~300
// records that nobody will ever read and that would bury the one line saying an account was deleted.
// Mutations are `info`. A refusal, a foreign-row reach and a failed sign-in are `warn` — those are the
// records somebody comes looking for. A 5xx is `error`.

import (
	"bufio"
	"net"
	"net/http"
	"strconv"
	"strings"
	"time"

	"github.com/AlexGromer/sentinel/internal/eventlog"
	"github.com/AlexGromer/sentinel/internal/svclog"
)

// statusRecorder captures the status code the handler wrote. net/http gives no way to read it back,
// and a journal that records every call as 200 would be worse than none — it would look complete.
type statusRecorder struct {
	http.ResponseWriter
	status int
	// foreign is set by mayTouch when it answers for a row belonging to somebody else. It rides here
	// because the refusal is written by mayTouch, several frames below the journal, and the alternative
	// — inferring "foreign" from a 404 — would file every genuinely missing row as an access attempt.
	foreign bool
}

func (w *statusRecorder) WriteHeader(code int) {
	if w.status == 0 {
		w.status = code
	}
	w.ResponseWriter.WriteHeader(code)
}

func (w *statusRecorder) Write(b []byte) (int, error) {
	if w.status == 0 {
		w.status = http.StatusOK // an unstamped write is a 200, exactly as net/http decides it
	}
	return w.ResponseWriter.Write(b)
}

// Flush and Hijack forward the two OPTIONAL interfaces net/http handlers reach for by type assertion.
// A wrapper that implements only ResponseWriter silently removes them, and both failures are ugly:
//
//	Flusher  — /v1/stream and /v1/live/mjpeg would arrive all at once at the end instead of streaming,
//	           a break no status code reveals.
//	Hijacker — /v1/stream's WebSocket upgrade (ws.go writes the 101 on a hijacked conn) answers 500.
//
// Measured, not anticipated: the first version forwarded Flush alone and TestStreamHandshakeAndIngest
// went red on the handshake. The gate below now asserts the SET of interfaces rather than one of them,
// because the next wrapper will forget a different one.
func (w *statusRecorder) Flush() {
	if f, ok := w.ResponseWriter.(http.Flusher); ok {
		f.Flush()
	}
}

func (w *statusRecorder) Hijack() (net.Conn, *bufio.ReadWriter, error) {
	h, ok := w.ResponseWriter.(http.Hijacker)
	if !ok {
		return nil, nil, http.ErrNotSupported
	}
	// A hijacked connection leaves net/http's status tracking behind entirely: the handler writes the
	// 101 itself. Recording it here keeps the journal honest about a call it can no longer observe.
	if w.status == 0 {
		w.status = http.StatusSwitchingProtocols
	}
	return h.Hijack()
}

// actorOf names WHO made a call, as a person reads it. Deliberately not the user id: an id answers
// "which row" and this field answers "who", and the two are different questions in a journal meant to
// be read rather than joined.
func (s *server) actorOf(r *http.Request) (actor, owner string) {
	c, ok := s.callerOf(r)
	switch {
	case !ok:
		return "anonymous", ""
	case c.machine:
		return "machine", ""
	}
	name := c.name
	if name == "" {
		name = c.owner()
	}
	if c.admin {
		return name + " (admin)", c.owner()
	}
	return name, c.owner()
}

// journalCall records one API call. Called from guard for EVERY route, after the handler ran.
func (s *server) journalCall(sp routeSpec, rec *statusRecorder, r *http.Request, started time.Time) {
	if s.journal == nil {
		return
	}
	actor, owner := s.actorOf(r)
	status := rec.status
	if status == 0 {
		status = http.StatusOK
	}
	dur := time.Since(started).Milliseconds()
	// The route PATTERN, never the concrete path: `GET /v1/runs/{id}` groups, and a journal of a
	// thousand distinct paths cannot be read by eye or counted by code. The id is not lost — it is in
	// the message, where it belongs to the sentence rather than to the index.
	route := sp.pattern
	if i := strings.IndexByte(route, ' '); i >= 0 {
		route = route[i+1:]
	}

	// A foreign-row touch is its own event, at its own level, because it is the one read anybody ever
	// comes looking for. It is NOT inferred from the 404 — mayTouch says so, so a genuinely missing
	// row is not filed as an access attempt.
	if rec.foreign {
		s.journal.Log(svclog.Record{
			Lvl: "warn", Cat: "service", Code: "service.foreign_row",
			Msg: s.renderMsg("service.foreign_row", map[string]string{
				"actor": actor, "method": r.Method, "route": route,
			}),
			Actor: actor, Owner: owner, Method: r.Method, Route: route,
			Status: status, DurMs: dur, Foreign: true,
		})
		return
	}

	code, lvl := "service.api_call", "debug"
	if r.Method != http.MethodGet && r.Method != http.MethodHead {
		lvl = "info" // a mutation is what a journal exists for
	}
	// A PROBE's non-2xx is its answer, not a fault (routeSpec.probe). /readyz replies 503 for as long
	// as a dependency is unconfigured, which for a standalone deployment is forever and is correct; an
	// orchestrator polling it would otherwise write an `error` record every few seconds and bury the
	// journal under the health of a healthy service.
	if !sp.probe {
		switch {
		case status >= 500:
			code, lvl = "service.api_refused", "error"
		case status == http.StatusForbidden || status == http.StatusUnauthorized:
			code, lvl = "service.api_refused", "warn"
		}
	}
	// The word REFUSED is in the sentence, not only in the code, and it no longer has to be spliced in
	// here: the two codes carry two templates, and the sentence comes from whichever one was chosen.
	s.journal.Log(svclog.Record{
		Lvl: lvl, Cat: "service", Code: code,
		Msg: s.renderMsg(code, map[string]string{
			"method": r.Method, "route": route,
			"status": strconv.Itoa(status), "dur_ms": strconv.FormatInt(dur, 10), "actor": actor,
		}),
		Actor: actor, Owner: owner, Method: r.Method, Route: route, Status: status, DurMs: dur,
	})
}

// renderMsg is the ONE place a service-journal sentence is produced (ADR-117). It renders the
// catalogue's template for `code`; a code the catalogue does not know gets the honest
// `eventlog.uncatalogued` line instead of an invented sentence, exactly as brain/eventlog.py does —
// because a code with no entry is a code the reader's browser cannot render either.
//
// ⚠ The code LITERALS stay at the call sites, here in cmd/control-api. tests/test_event_catalog_offline.py
// maps emitters to paths and finds codes by regexing them inside those paths; a literal that moved
// into the shared renderer would become a phantom the catalogue gate can no longer see.
func (s *server) renderMsg(code string, fields map[string]string) string {
	msg, ok := eventlog.Render(code, fields)
	if !ok {
		// Not silent, and not a guess: the same signal the Python side raises, so an emitter that
		// invents a code is visible in the very journal it was trying to write to.
		return "eventlog.uncatalogued: " + code + " is not in the catalogue"
	}
	return msg
}

// --- the messages emitted from main() ------------------------------------------------------------
//
// Functions rather than expressions assembled at the call site, and not as a style preference. The
// wording gate (svcjournal_wording_test.go) compares every emitted message against the catalogue's
// template for its code, because the catalogue's template is what the UI uses to render the sentence
// in the reader's language: when the two disagree, the extraction fails and a Russian reader silently
// gets English. A message built inline in main() cannot be reached from a test — which is how four of
// these came to disagree with the catalogue with every gate green, and stayed that way until somebody
// looked at a screenshot.

// ⚠ These four no longer ASSEMBLE anything — they name their fields and hand them to the catalogue's
// template. The functions remain (rather than the call sites calling Render directly) because the
// tests reach them, and because a named function is where the field NAMES for a code are decided
// once instead of at each caller.
func renderOrUncatalogued(code string, fields map[string]string) string {
	msg, ok := eventlog.Render(code, fields)
	if !ok {
		return "eventlog.uncatalogued: " + code + " is not in the catalogue"
	}
	return msg
}

func startedMsg(version, supervisor string, pid int, detail string) string {
	return renderOrUncatalogued("service.started", map[string]string{
		"svc": "control-api", "version": version, "supervisor": supervisor,
		"pid": strconv.Itoa(pid), "detail": detail,
	})
}

func stoppedMsg(reason string) string {
	return renderOrUncatalogued("service.stopped", map[string]string{"svc": "control-api", "reason": reason})
}

func tokenSourceMsg(source string, warnings int) string {
	return renderOrUncatalogued("service.token_source", map[string]string{
		"source": source, "detail": " — warnings: " + strconv.Itoa(warnings),
	})
}

func storeUnreachableMsg(addr, reason string) string {
	return renderOrUncatalogued("service.store_unreachable", map[string]string{
		"addr": addr, "reason": reason,
	})
}

// journalEvent records a service-plane event that is not an API call — a sign-in, a config write, an
// account change, the service coming up. `fields` are appended to the message rather than being typed
// columns: the catalogue's sentence is what a person reads, and every one of these is rare enough
// that a query over it would be a query over dozens of rows, not thousands.
func (s *server) journalEvent(code, lvl string, fields map[string]string, r *http.Request, extra ...string) {
	s.journalSubject(code, lvl, fields, r, "", extra...)
}

// journalSubject records an event ABOUT a particular account, for the cases where the request cannot
// name it. There are two, and both were recorded ownerless until a live check found them:
//
//	a successful SIGN-IN happens before the session exists, so actorOf(r) sees an anonymous caller;
//	an account CREATED or DELETED by an admin is not the caller at all — the admin is.
//
// Without a subject those records carry no owner, and an ownerless service event is admin-only
// (svcjournal_read.go). That makes the read route's whole reason for being `authed` rather than
// `admin` — an account reading the record of its own sign-ins — not work. The unit fixture did not
// catch it because the fixture SET the owner it then asserted on; the product never did.
// ⚠ THE SENTENCE IS NOT A PARAMETER (ADR-117). It used to be, and that is precisely how six codes
// came to disagree with the catalogue templates the hub matches them against. Now the caller names
// FIELDS and the template does the rest, so a hand-assembled sentence is not merely discouraged —
// there is nowhere to pass one.
//
// `extra` keeps its old shape and its old job, but it now feeds the template's `{detail}` field
// instead of being appended after rendering. That distinction is the whole point: appended text sits
// OUTSIDE the template, so the hub's extractor has to absorb it into whatever placeholder happens to
// be last, and a template edit silently changes what "the detail" means.
func (s *server) journalSubject(code, lvl string, fields map[string]string, r *http.Request, subject string, extra ...string) {
	if s.journal == nil {
		return
	}
	var actor, owner string
	if r != nil {
		actor, owner = s.actorOf(r)
	}
	if subject != "" {
		// The SUBJECT wins over the caller: an admin deleting somebody's account is actor=admin,
		// owner=the other account, which is exactly the case collapsing the two would lose.
		owner = subject
	}
	if len(extra) > 0 {
		if fields == nil {
			fields = map[string]string{}
		}
		fields["detail"] = fields["detail"] + " — " + strings.Join(extra, ", ")
	}
	s.journal.Log(svclog.Record{
		Lvl: lvl, Cat: "service", Code: code, Msg: s.renderMsg(code, fields), Actor: actor, Owner: owner,
	})
}

File: pw-executor/src/svcjournal.ts
/**
 * The SERVICE journal, written from the browser service (HEALTH-005 PR-C, ADR-116).
 *
 * WHY A THIRD WRITER. The journal answers "what did the tool itself do" — sign-ins, configuration
 * changes, services starting and stopping — and until now the browser service was absent from it
 * entirely. It logged with `console.error`, which means the container's stdio: not a file, not
 * catalogued, not filterable, not in the UI, and destroyed by `docker compose down`. A service that
 * is part of the default stack and leaves no trace in the deployment's own record is a hole in
 * exactly the thing the record exists for.
 *
 * SAME FILE, NOT A FOURTH ONE. `state/logs/service.jsonl`, the same one control-api and agentctl
 * write, distinguished by `svc`. Four files would mean answering "what happened at 14:32" by merging
 * them by hand — the reason ADR-116 chose one file in the first place.
 *
 * THREE PROPERTIES THIS DELIBERATELY DOES NOT HAVE, each for a stated reason:
 *
 *   No rotation. Two rotators racing on one file is worse than one: whoever renames second discards
 *   the generation the first had just created. The Go writer sizes the file by SEEKING it rather
 *   than by a counter (svclog.Log, PR-B), so it rotates on the real size no matter who wrote the
 *   bytes — and this service writes about two records per lifetime, so it contributes almost none.
 *
 *   No lock. `flock` needs a native module and this project takes no new runtime dependencies for
 *   it. The consequence is real and is NOT hidden: a record appended here while `agentctl
 *   purge-service` rewrites the file could be lost. That window is what the purge's carry-over pass
 *   exists to close (internal/svclog/purge.go), and the residual gap is declared in
 *   docs/OBSERVABILITY.md rather than implied away.
 *
 *   No throwing. A service must not fail to start over its own log file — the same rule that makes
 *   `svclog.Open` return nil instead of an error. Every failure here is swallowed after one line on
 *   stderr, because a logging problem must not become an outage.
 */
import * as fs from 'node:fs';
import * as path from 'node:path';

/** One line of the journal. The field names are the wire format — `svclog.Record`'s JSON tags. */
export interface JournalRecord {
  seq: number;
  ts: string;
  lvl: string;
  cat: string;
  code: string;
  msg: string;
  svc: string;
}

/** Where the journal lives. The compose services mount ./state at /app/state. */
export const stateDir = (): string => process.env.SENTINEL_STATE_DIR ?? '/app/state';

let seq = 0;
let reported = false;

/**
 * Which supervisor brought this process up.
 *
 * ⚠ This mirrors `svclog.Supervisor()` (Go) and is therefore a SECOND implementation of one rule —
 * the shape this milestone spent a PR removing elsewhere. It is here because the two live on
 * opposite sides of a language boundary with no shared runtime, and the alternative is worse: either
 * omit the field the catalogue's template names, or fill it with a guess. Kept to the same signals
 * as the Go original, deliberately in the same order, so a reader comparing them can see they agree.
 * Recorded in the backlog under [DEBT-GO-EVENTLOG].
 */
export function supervisor(): string {
  if (process.env.INVOCATION_ID) return 'systemd';
  try {
    if (fs.existsSync('/.dockerenv')) return 'container';
  } catch { /* an unreadable filesystem is not a reason to guess */ }
  return 'manual';
}

/**
 * The name THIS process answers to in the deployment's record — the one place it is decided.
 *
 * WHY IT IS A FUNCTION AND NOT A LITERAL AT EACH USE. Two services run this same binary: `browser`
 * (headless, the default stack) and `browser-vnc` (headed, behind the `vnc` profile, ADR-127). They
 * append to the SAME journal file, so this value is the only thing that tells their lines apart.
 * Until now it was read here for the `svc` FIELD and written as the literal `browser` in the message
 * TEXT — two statements about one fact, and they disagreed the first time anybody measured them on a
 * live stack: `{"code":"service.started","msg":"Service browser started: version dev, …",
 * "svc":"browser-vnc"}` (docker, 2026-08-18). The field was right and the sentence was wrong, which
 * is the worse way round: the hub's "Service journal" view shows people the TEXT, so two services
 * read as one and the difference survives only in the machine representation.
 *
 * ⚠ WHY NO GATE SAW IT, and what that demands of the new ones. The catalogue template is
 * `Service {svc} started: …` and `{svc}` accepts ANY value, so matching a message against its
 * template — the check that exists precisely for this class — matched the WRONG name perfectly.
 * A test that only renders this service's own default name is vacuous for the same reason: the
 * literal and the correct answer are the same string. Hence svcjournal.test.ts drives a name this
 * process is NOT running under, and asserts the sentence and the `svc` field agree.
 */
export function svcName(): string {
  return process.env.SENTINEL_SVC_NAME || 'browser';
}

/**
 * The `service.started` sentence, built to match its catalogue template exactly.
 *
 * Exported, and built here rather than at the call site, for the reason PR-B measured the hard way:
 * the hub renders a record in the reader's language by matching the catalogue's English template
 * against the message the service actually sent, so a message that drifts from its template falls
 * back to raw English — silently, one row at a time. Six Go codes were doing that. A message
 * assembled inline cannot be reached by a test; this one can, and tests/test_browser_journal_offline.py
 * runs THIS function and matches its output against brain/events.json.
 *
 * ⚠ `svc` is LAST and defaults to svcName() on purpose. The shipping call site passes nothing, so the
 * name in the sentence and the name in the record's `svc` field come from ONE function call apiece to
 * ONE function — they cannot drift without changing the line they both read, which is the whole point
 * of the parameter. It is a parameter at all, rather than an unconditional svcName() inside, because
 * a test that cannot say a name other than this process's own cannot tell a derived name from the
 * hardcoded one it replaced.
 */
export function startedMsg(version: string, sup: string, pid: number, detail: string,
                           svc: string = svcName()): string {
  return `Service ${svc} started: version ${version}, brought up by ${sup}, pid ${pid}${detail}`;
}

/** The `service.stopped` sentence, same contract as startedMsg — including where `svc` comes from. */
export function stoppedMsg(reason: string, svc: string = svcName()): string {
  return `Service ${svc} stopped: ${reason}`;
}

/**
 * The `service.live_claim_conflict` sentence — two runs declaring ONE page (ADR-128).
 *
 * WHY A RARE EVENT GETS A CODE RATHER THAN A LOG LINE. Until ADR-128 this was the topology: in
 * CDP-attach mode every run adopted `pages()[0]`, so a second concurrent run announced the first
 * one's target by construction (measured live: both `84DC6185`). A run now opens its own page, which
 * makes the collision UNREACHABLE in a deployment whose parts agree — and that is exactly what turns
 * it from noise into a signal. If it fires now, the executor driving this service is older than
 * ADR-128, and the symptom a person will actually meet is "the live view refuses both of my runs"
 * with the cause on stderr of a container nobody is tailing. A code puts it in the deployment's own
 * record, where the hub, the CLI and the API can all read it.
 */
export function claimConflictMsg(runs: string, target: string): string {
  return `Runs ${runs} claimed one browser page ${target} — the live view cannot say which of them it shows`;
}

/** Append one record. Never throws; never blocks on anything but the write itself. */
export function journal(code: string, lvl: string, msg: string): void {
  const dir = path.join(stateDir(), 'logs');
  const line = JSON.stringify({
    seq: ++seq,
    ts: new Date().toISOString(),
    lvl,
    cat: 'service',
    code,
    msg,
    // LIVE-VNC: TWO services run this same binary and append to the SAME file, so this field is what
    // answers "which service said this". It reads svcName() rather than the environment directly
    // because the message builders above read the same function: the field and the sentence are one
    // fact, and reading the variable twice is how they came to disagree (see svcName).
    svc: svcName(),
  } satisfies JournalRecord) + '\n';
  try {
    // 0750/0640 match what the Go writer creates, so the file's permissions do not depend on which
    // service happened to create it first.
    fs.mkdirSync(dir, { recursive: true, mode: 0o750 });
    // ONE write of a complete line: below PIPE_BUF, so it cannot interleave with another appender's
    // line and produce a torn record. Synchronous on purpose — the stop record is written while the
    // process is on its way out, and an async write would not survive the exit.
    fs.appendFileSync(path.join(dir, 'service.jsonl'), line, { mode: 0o640 });
  } catch (e) {
    if (!reported) {
      reported = true;
      process.stderr.write(
        `[cdp-service] service journal write failed: ${(e as Error).message} (further failures are silent)\n`,
      );
    }
  }
}
